File: LeetCode/198. House Robber/Solution.kt
package problem198

import kotlin.math.max

// use brain power!!!!!!!!!
class Solution {
  fun rob(nums: IntArray): Int {
    if (nums.size == 1) {
      return nums[0]
    }

    // total money if nums[i] must be taken
    val moneyTake = IntArray(nums.size)
    moneyTake[0] = nums[0]

    for (i in 1..<nums.size) {
      // [x][x][x][i]
      //     v     v <- solution 1
      //  v        v <- solution 2
      // see which is bigger
      var money = nums[i]
      if (i - 2 >= 0) {
        money = max(money, moneyTake[i - 2] + nums[i])
      }
      if (i - 3 >= 0) {
        money = max(money, moneyTake[i - 3] + nums[i]
        )
      }
      moneyTake[i] = money
    }

    return max(moneyTake[nums.size - 1], moneyTake[nums.size - 2])
  }
}

// test
fun test(
  // input
  vararg input: Int,

  // expected
  expected: Int,
) {

  val result = Solution().rob(input)
  println("${input.contentToString()} -> ${result} == ${expected} -> ${result == expected}")
}

fun main() {
  test(2, 1, 1, 2, expected = 4)
  test(1, 2, 3, 1, expected = 4)
  test(2, 3, 2, expected = 4)
  test(2, 4, 3, expected = 5)
  test(2, 4, 3, 2, expected = 6)
  test(1, 2, 3, 1, expected = 4)
  test(2, 7, 9, 3, 1, expected = 12)
}

File: LeetCode/198. House Robber/Solution.cc
#include <algorithm>  // for std::min
#include <vector>

using std::vector;

class Solution {
 public:
  int rob(const vector<int>& nums) {
    if (nums.size() == 1) {
      return nums[0];
    }

    vector<int> money(nums.size());
    money[0] = nums[0];

    // [x][x][x][i]
    //     v     v <- solution 1
    //  v        v <- solution 2
    // see which is bigger
    for (size_t i = 1; i < nums.size(); i++) {
      int m = nums[i];
      if (i >= 2) {
        m = std::max(m, money[i - 2] + nums[i]);
      }
      if (i >= 3) {
        m = std::max(m, money[i - 3] + nums[i]);
      }
      money[i] = m;
    }

    return std::max(money[nums.size() - 1], money[nums.size() - 2]);
  }
};

// test

#include <initializer_list>
#include <iostream>

void test(std::initializer_list<int> input, int expected) {
  int ret = Solution().rob(std::vector<int>{input});
  std::cout << ret << " == " << expected << std::endl;
}

int main() {
  test({2, 1, 1, 2}, 4);
  test({1, 2, 3, 1}, 4);
  test({2, 3, 2}, 4);
  test({2, 4, 3}, 5);
  test({2, 4, 3, 2}, 6);
  test({1, 2, 3, 1}, 4);
  test({2, 7, 9, 3, 1}, 12);
}
